pkg/telemetry/span.go

```diff
@@ -0,0 +1,439 @@
+
+// End utilizes the internal tracer's processors to send spans
+func (s *Span) End(opts ...trace.SpanEndOption) {
+	s.EndedAt = time.Now()
+
+	if err := UserTracer().Export(s); err != nil {
+		ctx := context.Background()
+		log.From(ctx).Error().Err(err).Msg("error ending span")
+	}
+}
+
+func (s *Span) AddEvent(name string, opts ...trace.EventOption) {
+	s.Lock()
+	defer s.Unlock()
+
+	config := trace.NewEventConfig(opts...)
+
+	evt := tracesdk.Event{
+		Name:       name,
+		Time:       time.Now(),
+		Attributes: config.Attributes(),
+	}
+	if !config.Timestamp().IsZero() {
+		evt.Time = config.Timestamp()
+	}
+
+	s.SpanEvents = append(s.SpanEvents, evt)
+}
+
+func (s *Span) IsRecording() bool {
+	return true
+}
+
+// official one doesn't actually set the status, but we'll just do it here
+// for convinence's sake.
+func (s *Span) RecordError(err error, opts ...trace.EventOption) {
+	s.Lock()
+	defer s.Unlock()
+
+	s.AddEvent(err.Error(), opts...)
+	s.SetStatus(codes.Error, err.Error())
+}
+
+func (s *Span) SetStatus(code codes.Code, desc string) {
+	s.Lock()
+	defer s.Unlock()
+	s.SpanStatus = tracesdk.Status{
+		Code:        code,
+		Description: desc,
+	}
+}
+
+func (s *Span) SetName(name string) {
+	s.Lock()
+	defer s.Unlock()
+	s.SpanName = name
+}
+
+// SetAttributes mimics the official SetAttributes method, but with
+// reduced checks. We're not doing crazy stuff with it so there's
+// less of a need to do so.
+func (s *Span) SetAttributes(attrs ...attribute.KeyValue) {
+	s.Lock()
+	defer s.Unlock()
+
+	if s.Attrs == nil {
+		s.Attrs = []attribute.KeyValue{}
+	}
+
+	// dedup if the sum of existing and new attr could exceed limit
+	if len(s.Attrs)+len(attrs) > attrCountLimit {
+		// dedup the existing list of attributes and take the latest one
+		exists := make(map[attribute.Key]int)
+		dedup := []attribute.KeyValue{}
+		for _, a := range s.Attrs {
+			if idx, ok := exists[a.Key]; ok {
+				dedup[idx] = a
+			} else {
+				dedup = append(dedup, a)
+				exists[a.Key] = len(dedup) - 1
+			}
+		}
+
+		for _, a := range attrs {
+			if !a.Valid() {
+				// Drop invalid attributes
+				s.droppedAttributes++
+				continue
+			}
+
+			// if a key is already there, take the latest one
+			if idx, ok := exists[a.Key]; ok {
+				s.Attrs[idx] = a
+				continue
+			}
+
+			// don't bother appending if it's at limits
+			if len(s.Attrs) >= attrCountLimit {
+				s.droppedAttributes++
+				continue
+			}
+
+			s.Attrs = append(s.Attrs, a)
+			exists[a.Key] = len(s.Attrs) - 1
+		}
+	}
+
+	// otherwise, just append
+	for _, a := range attrs {
+		if !a.Valid() {
+			s.droppedAttributes++
+			continue
+		}
+		s.Attrs = append(s.Attrs, a)
+	}
+}
+
+func (s *Span) TracerProvider() trace.TracerProvider {
+	return UserTracer().Provider()
+}
+
+func newSpanIDGenerator() *spanIDGenerator {
+	var seed int64
+	_ = binary.Read(crand.Reader, binary.LittleEndian, &seed)
+
+	return &spanIDGenerator{
+		randSrc: rand.New(rand.NewSource(seed)),
```

Comment:
This is irrelevant. It follows the similar implementation as the official otel ID generator. I'll dismiss it when the PR is approved and ready to merge, otherwise it keeps popping up.